document: add contributing guide (#132)
pingcap-incubator/tidb-dashboard: CONTRIBUTING.md

- `make` - Build tool (run common workflows)
- [`Golang`](https://golang.org/) - Golang (require 1.13+)
- [`Node.js`](https://nodejs.org/) - Javascript runtime (require 12+)
- [`Yarn`](https://classic.yarnpkg.com/en/docs/install) - Javascript dependency management (require 1.21+)
- [`Java`](https://www.java.com/ES/download/) - depended by [openapi-generator](https://github.com/OpenAPITools/openapi-generator) (require 8+)

### Getting the repository

```bash
git clone https://github.com/pingcap-incubator/tidb-dashboard.git
cd tidb-dashboard
# Future instructions assume you are in this directory
```

Optional: set Taobao's npm mirror registry to speed up package downloading if needed

```bash
npm config set registry https://registry.npm.taobao.org
```

### Building and running

At this point, you can build and run TiDB Dashboard. 

> Note: If you want to debug TiDB Dashboard, it needs a running TiDB cluster as target, see [how to start a local TiDB cluster](#starting-a-tidb-cluster).

Build and run back-end server:

```bash
make
make run
# Now tidb-dashboard server is listen on 127.0.0.1:12333
```

For front-end, you should build API client and start a React development server:

> ```bash

Reviewer: do we need to add `yarn`?

cd ui
npm run build_api_client
npm run start
# Now tidb-dashboard UI is available on 127.0.0.1:3000
```

When you're ready to test out your changes, use the `dev` task. It will lint your code and verify the UI building.

```bash

> make dev

Reviewer: this part only applies to back-end develop. Maybe better to put it before `For front-end, ...`

Author: As we will add `js linter` in the future, I think we can keep content here and add front-end check command here in the future.

```

See the [style doc](https://github.com/golang/go/wiki/CodeReviewComments) for details on the conventions.

Please follow this style to make TiDB Dashboard easy to review, maintain, and develop.

### Starting a TiDB cluster

> > Note: TiDB Dashboard can run without a TiDB cluster, but some features are unavailable and not conducive to debugging.

Reviewer: This note can be removed. Without a TiDB cluster login is even not possible now..

> To run TiDB Dashboard, you'd like to run a local TiDB cluster (at least 1 TiDB, 1 TiKV and 1 PD), here we introduce how to start a local TiDB cluster by binary deployment. Alternatively, you can use other ways like [docker-compose](https://github.com/pingcap/tidb-docker-compose#quick-start) or [TiUP](https://tiup.io) (It's under actively development).

Reviewer: TiDB Dashboard does not work well with docker deployed cluster now. I think we can remove it.

Download all needed files (Do not save them in `tidb-dashboard` folder):

```bash
# Download the package.

> wget https://download.pingcap.org/tidb-latest-linux-amd64.tar.gz

Reviewer: This only covers linux instructions. The commonly used MacOS platform is not covered.

```

For front-end, you should build API client and start a React development server:

```bash
cd ui
yarn # Install all the dependencies
npm run build_api_client
npm run start
# Now tidb-dashboard UI is available on 127.0.0.1:3000
```

When you're ready to test out your changes, use the `dev` task. It will lint your code and verify the UI building.

```bash
make dev
```

See the [style doc](https://github.com/golang/go/wiki/CodeReviewComments) for details on the conventions.

Please follow this style to make TiDB Dashboard easy to review, maintain, and develop.

### Starting a TiDB cluster

To run TiDB Dashboard, you'd like to run a local TiDB cluster (at least 1 TiDB, 1 TiKV and 1 PD), here we introduce how to start a local TiDB cluster by binary deployment. Alternatively, you can use other ways like [TiUP](https://tiup.io) (It's under actively development).

Download all needed files, but do not save them in `tidb-dashboard` folder, Choose 1 of 2:

For linux user:
```bash
wget https://download.pingcap.org/tidb-latest-linux-amd64.tar.gz
tar -xzf tidb-latest-linux-amd64.tar.gz
cd tidb-latest-linux-amd64
```

For Mac OS user: 
```bash

> wget https://download.pingcap.org/tidb-latest-darwin-amd64.tar.gz

Reviewer 2: Only `tidb-server` can be obtained.

# Contributing to TiDB Dashboard

Thanks for your interest in contributing to TiDB Dashboard! This document outlines some of the conventions on building, running, and testing TiDB Dashboard, the development workflow, commit message formatting, contact points and other resources.

> If you need any help or mentoring getting started, understanding the codebase, or making a PR (or anything else really), please ask in `#sig-dashboard` channel on [Slack](https://join.slack.com/t/tidbcommunity/shared_invite/enQtNzc0MzI4ODExMDc4LWYwYmIzMjZkYzJiNDUxMmZlN2FiMGJkZjAyMzQ5NGU0NGY0NzI3NTYwMjAyNGQ1N2I2ZjAxNzc1OGUwYWM0NzE).

Reviewer: Better to post a Chinese channel as well (mark Chinese).

## Building and setting up a development workspace

> TiDB Dasboard is a web interface which integrate into PD by default, andd can be deploy standalone. Its back-end is written in [Golang](https://golang.org/) with [Gin](https://github.com/gin-gonic/gin), front-end is written in [Typescript](https://www.typescriptlang.org/) with [React](https://github.com/facebook/react). It uses [swaggo](https://github.com/swaggo/swag) and [openapi-generator](https://github.com/OpenAPITools/openapi-generator) for automatically generating API documents and API client. Also, to provide consistency, we use linters and automated formatting tools.

Reviewer 3: ```suggestion
TiDB Dashboard is a web interface that integrates into PD by default and can be deployed standalone. Its back-end is written in [Golang](https://golang.org/) with [Gin](https://github.com/gin-gonic/gin), front-end is written in [Typescript](https://www.typescriptlang.org/) with [React](https://github.com/facebook/react). It uses [swaggo](https://github.com/swaggo/swag) and [openapi-generator](https://github.com/OpenAPITools/openapi-generator) for automatically generating API documents and API client. Also, to provide consistency, we use linters and automated formatting tools.
```

> TiDB Dasboard is a web interface which integrate into PD by default, andd can be deploy standalone. Its back-end is written in [Golang](https://golang.org/) with [Gin](https://github.com/gin-gonic/gin), front-end is written in [Typescript](https://www.typescriptlang.org/) with [React](https://github.com/facebook/react). It uses [swaggo](https://github.com/swaggo/swag) and [openapi-generator](https://github.com/OpenAPITools/openapi-generator) for automatically generating API documents and API client. Also, to provide consistency, we use linters and automated formatting tools.

Reviewer: ```suggestion
TiDB Dashboard is a web based UI for TiDB clusters. It is integrated into [PD](https://github.com/pingcap/pd) by default. It can also be deployed as a standalone server. The back-end is written in [Golang](https://golang.org/) with [Gin web framework](https://github.com/gin-gonic/gin) and the front-end is written in [TypeScript](https://www.typescriptlang.org/) with [React](https://github.com/facebook/react). [swaggo](https://github.com/swaggo/swag) and [openapi-generator](https://github.com/OpenAPITools/openapi-generator) are used for automatically generating API documents and API client.
```

> 

Reviewer: We may need to explain that the following steps are describing how to develop TiDB Dashboard by running a self-built TiDB Dashboard server (i.e. standalone) along with a separated TiDB, TiKV, PD cluster.

### Prerequisites

To build TiDB Dashboard you'll need to at least have the following installed:

- `git` - Version control
- `make` - Build tool (run common workflows)

> - [`golang`](https://golang.org/) - Golang (require 1.13+)

Reviewer: Golang

> - [`node.js`](https://nodejs.org/) - Javascript runtime (require 12+)

Reviewer 2: I think yarn needs to be added here too.

Reviewer: +1. Otherwise yarn.lock will be incorrect.

> - [`node.js`](https://nodejs.org/) - Javascript runtime (require 12+)

Reviewer: Node.js

Reviewer: As reported by @HunDunDM , Java is also needed for running the openapi-generator.

### Getting the repository

```bash
git clone https://github.com/pingcap-incubator/tidb-dashboard.git
cd tidb-dashboard
# Future instructions assume you are in this directory
```

### Preparing environment

Optional: set Taobao's npm mirror registry to speed up package downloading if needed

```bash
npm config set registry https://registry.npm.taobao.org
```

Install yarn

```bash

> npm install yarn -g

Reviewer 2: This is not an officially recommended installation method.

Reviewer 4: Yep, maybe  we can just attach an official link here: https://classic.yarnpkg.com/en/docs/install

```

> Install Openapi-generator

Reviewer 2: openapi-generator-cli is a dependency of the project and does not need to be installed globally.

```bash
npm install @openapitools/openapi-generator-cli -g
```

**IMPORTANT**: TiDB Dashboard uses packages from GitHub Packages. you need to configure the ~/.npmrc in order to install these packages.

1. Create a personal access token

    Follow personal ["Settings" -> "Developer settings" -> "Personal access tokens" -> "Generate new token"](https://github.com/settings/tokens/new) path, the "Select scopes" must select `repo` and `read:packages` at least.

2. Edit `~/.npmrc`, add a new line by the following content

```
//npm.pkg.github.com/:_authToken=[YOUR_PERSONAL_ACCESS_TOKEN]
```

### Building and running

At this point, you can build and run TiDB Dashboard. 

> Note: TiDB Dashboard need a running TiDB cluster as target, before continue, see [how to start a local TiDB cluster](#starting-a-tidb-cluster).

Build and run back-end server:

```bash
make
make run
# Now tidb-dashboard server is listen on 127.0.0.1:12333
```

> For front-end, you should build API client and start a React develpment server:

Reviewer 3: ```suggestion
For front-end, you should build API client and start a React development server:
```

```bash
make swagger_client
cd ui

> npm run start

Reviewer: I prefer:

```
cd ui
npm run build_api_client
npm run start
```

# Now tidb-dashboard UI is available on 127.0.0.1:3000
```

When you're ready to test out your changes, use the `dev` task. It will lint your code and verify the UI building.

```bash
make dev
```

See the [style doc](https://github.com/golang/go/wiki/CodeReviewComments) for details on the conventions.

Please follow this style to make TiDB Dashboard easy to review, maintain, and develop.

> ### Starting a TiDB cluster

Reviewer 3: Can we make start TiDB as an option?

Author: Yes, I clarify this in doc.

> To run TiDB Dashboard, you also need run a local TiDB cluster (at least 1 TiDB server), here we introduce how to start a local TiDB cluster by binary deployment. Also, you can use other ways like [docker-compose](https://github.com/pingcap/tidb-docker-compose#quick-start) or [TiUP](https://tiup.io).

Reviewer: I think we can put more texts for TiUP. It is much more easy to use TiUP to bootstrap a cluster locally.

Author: TiUP is not ready now, We use it as default way in the future.

Reviewer: Its playground feature is enough for our development

Download all needed files (Do not save them in `tidb-dashboard` folder):

```bash
# Download the package.
wget https://download.pingcap.org/tidb-latest-linux-amd64.tar.gz
# Extract the package.
tar -xzf tidb-latest-linux-amd64.tar.gz
cd tidb-latest-linux-amd64
```

> You can simply start 1 TiDB server for testing:

Reviewer 2: Need to explain that we cannot use the dashboard when there is only the tidb node?

Author: ok, I clarify this in doc.

For beginners, we have prepared many suitable tasks for you. Checkout our [Help Wanted issues](https://github.com/pingcap-incubator/tidb-dashboard/issues?q=is%3Aopen+label%3Astatus%2Fhelp-wanted+sort%3Aupdated-desc) for a list, in which we have also marked the difficulty level.

If you are planning something big, for example, relates to multiple components or changes current behaviors, make sure to open an issue to discuss with us before going on.

### Format of the commit message

We follow a rough convention for commit messages that is designed to answer two
questions: what changed and why. The subject line should feature the what and
the body of the commit should describe the why.

    clusterinfo: add comment for variable declaration.
    
    Improve documentation.

The format can be described more formally as follows:

    <subsystem>: <what changed>
    <BLANK LINE>
    <why this change was made>
    <BLANK LINE>
    Signed-off-by: <Name> <email address>

The first line is the subject and should be no longer than 50 characters, the other lines should be wrapped at 72 characters (see [this blog post](https://preslav.me/2015/02/21/what-s-with-the-50-72-rule/) for why).

> If the change affects more than one subsystem, you can use comma to separate them like `util/codec,util/types:`.

Reviewer 3: It's better to change it to something related to this repo.

If the change affects many subsystems, you can use `*` instead, like `*:`.

The body of the commit message should describe why the change was made and at a high level, how the code works.

### Signing off the Commit

The project uses [DCO check](https://github.com/probot/dco#how-it-works) and the commit message must contain a `Signed-off-by` line for [Developer Certificate of Origin](https://developercertificate.org/).

> Use option `git commit -s` to sign off your commits.

Reviewer: We don't use DCO check. This can be removed.
